src/cart/src/cart/crt_group.c

```diff
@@ -341,9 +315,17 @@ grp_li_uri_set(struct crt_lookup_item *li, int tag, const char *uri)
 	struct crt_uri_item	*ui;
 	d_list_t		*rlink;
 	struct crt_grp_priv	*grp_priv;
+	crt_phy_addr_t		uri_dup = NULL;
+	crt_phy_addr_t		nul_str = NULL;
 	d_rank_t		rank;
 	int			rc = 0;
 
+	D_STRNDUP(uri_dup, uri, CRT_ADDR_STR_MAX_LEN);
+	if (!uri_dup) {
+		D_ERROR("Failed to duplicate an uri\n");
+		D_GOTO(exit, rc = -DER_NOMEM);
+	}
+
```

Comment:
we shouldnt alloc/dup unless we don't have an entry or we have ui entry but its not populated for the tag. When we dont have a URI for the rank:tag each rpc will generate it's own lookup which will hit this function until one of them will populate it.
Applications that submit lots of rpcs without waiting for response will generate a lot of lookups with all of initial rpcs hitting this function at the same time. Doing allocs/free for all of those is not efficient

(As a note, once simpler synchronous hg lookup is landed in mercury, this will be greatly simplified)